main/src/library/Iterator.flix

```diff
@@ -589,199 +517,169 @@ namespace Iterator {
-    ///
-    def flatMapForward(f: a -> Iterator[b, r1] \ ef, done: Done[r2], next: Next[a, r2]): Option[Iterator[b, r1]] \ { ef, Read(r1, r2) } =
-        if (done())
-            None
-        else
-            let Iterator(d, n) = f(next());
-            if (d()) flatMapForward(f, done, next) else Some(Iterator(d, n))
+        let iter1F = () -> outerLoop();
+        Iterator(rc, iter1F)
 
     ///
     /// Returns an iterator with `a` inserted between every of `iter`.
     ///
     /// Does *not* consume any elements from the iterator.
     ///
-    @Lazy
-    pub def intersperse(x: a, iter: Iterator[a, r]): Iterator[a, r] \ Write(r) =
-        let Iterator(done, next) = iter;
-        let returnX = ref false @ Scoped.regionOf(iter);
-        let next1 = () -> {
-            if (deref returnX) {
-                returnX := false;
-                x
-            }
-            else {
-                returnX := true;
-                next()
-            }
-        };
-        Iterator(done, next1)
+    pub def intersperse(sep: a, iter: Iterator[a, ef, r]): Iterator[a, ef and r, r] \ Write(r) =
+        let Iterator(rc, _) = iter;
+        let start = ref true @ rc;
+        let step = x ->
+            if (deref start) {
+                start := false;
+                singleton(rc, x)
+            } else {
+                cons(sep, singleton(rc, x))
+            };
+        flatMap(step, iter)
 
     ///
-    /// Returns the concatenation of the elements in `iterB` with the elements of `iterA` inserted between every two adjacent elements.
+    /// Returns the concatenation of the elements in `iter` with the elements of `sep` inserted between every two adjacent elements.
     ///
     /// That is, returns `b1 :: a1 ... an :: b2 :: ... bn-1 :: a1 :: ... :: an :: bn :: Nil`.
     ///
     /// Does *not* consume any elements from either iterator.
     ///
-    /// The original iterators `iterA` and `iterB` should not be reused.
-    ///
-    @Lazy
-    pub def intercalate(iterA: Iterator[a, r1], iterB: Iterator[Iterator[a, r2], r3]): Iterator[a, r1 and r2 and r3] \ Write(r1, r3) =
-        let Iterator(doneA, nextA) = iterA;
-        let Iterator(doneB, nextB) = iterB;
-        let r1 = Scoped.regionOf(iterA);
-        let xs = ref Nil @ r1;
-        let r3 = Scoped.regionOf(iterB);
-        let isReversed = ref false @ r1;
-        let nextAA = () -> {
-            let x = nextA();
-            xs := x :: deref xs;
-            x
-        };
-        let doneAA = () -> {
-            if (doneA()) {
-                if (deref isReversed)
-                    ()
-                else {
-                    xs := List.reverse(deref xs);
-                    isReversed := true
-                };
-                true
+    /// The original iterators `sep` and `iter` should not be reused.
+    ///
+    pub def intercalate(sep: Iterator[a, ef2, r], iter: Iterator[Iterator[a, ef1, r], ef1, r]): Iterator[a, ef1 and ef2 and r, r] \ Write(r) =
```

Comment:
What's the meaning of this? Will `sep` not be consumed after it has been intercalated once?

Reply:
`sep` is consumed (turned into a List) the first time `step` is run. As a list it is stored in the ref cell `sepList` so a new "sep" iterator can be rebuilt at each gap between the iterator-of-iterators.

Reply:
I see. But then I find the function strange/unusual. 

Would it not be better if `sep` was foldable? Then you can iterate through it as often as you like.

Right now, if you have a list and an iterator, you have to get an iterator for the list to call intercalate.